### PR TITLE
Inviting users

### PR DESCRIPTION
**Pull request by jenmccarthy**
An admin has the ability to invite a student to use epicenter. Let me know about the testing coverage...I've tested the integration features, but I did not test the ability for an admin to create a student. The actual inviting and accepting an invite are done magically with devise_invitable and the docs say that a Rails app won't need to include the methods to handle that.  But that means I didn't test for the functionality of that. 

I've removed the ability for someone to visit the new_student_registration path (student/sign_up), but that route is the ugly action controller routing error page...is there a way to build in a better error view page for that? Or does it matter?

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -8,6 +8,10 @@ class ApplicationController < ActionController::Base
 
 protected
   def configure_permitted_params
+    devise_parameter_sanitizer.for(:accept_invitation) do |u|
+    u.permit(:name, :email, :cohort_id, :plan_id, :password, :password_confirmation,
```
Comment (michaelrkn) [inline app/controllers/application_controller.rb]: indent this line

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -8,6 +8,10 @@ class ApplicationController < ActionController::Base
...
+             :invitation_token)
+    end
     devise_parameter_sanitizer.for(:sign_up) { |u| u.permit(:plan_id, :cohort_id, :name, :email, :password, :password_confirmation) }
```
Comment (michaelrkn) [inline app/controllers/application_controller.rb]: break this onto multiple lines - too long!

File: app/models/student.rb [hunk 3]
```diff
@@ -1,4 +1,7 @@
 class Student < User
+
+  devise :invitable
```
Comment (michaelrkn) [inline app/models/student.rb]: if this is already on `User`, it shouldn't need to be on `Student` too, right?

File: app/views/devise/invitations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<h2><%= t 'devise.invitations.edit.header' %></h2>
```
Comment (michaelrkn) [inline app/views/devise/invitations/edit.html.erb]: add bootstrap styling

File: app/views/devise/mailer/invitation_instructions.html.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<p><%= I18n.t("devise.mailer.invitation_instructions.hello", email: @resource.email) %></p>
```
Comment (michaelrkn) [inline app/views/devise/mailer/invitation_instructions.html.erb]: We should probably customize this with something nicer in the message, like "We're so excited to have you join us at Epicodus! To finish your enrollment and pay your tuition, click here: [link]."

File: app/views/devise/registrations/edit.html.erb [hunk 6]
```diff
@@ -11,7 +11,7 @@
   </div>
 
   <div class='span9'>
-  <%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+  <%= form_for(resource, as: resource_name, url: user_registration_path, html: { method: :put }) do |f| %>
```
Comment (michaelrkn) [inline app/views/devise/registrations/edit.html.erb]: why is this changed?

File: spec/features/students_pages_spec.rb [hunk 7]
```diff
@@ -1,23 +1,25 @@
...
-    @plan = FactoryGirl.create :recurring_plan_with_upfront_payment
-    @cohort = FactoryGirl.create :future_cohort
-    visit new_student_registration_path
-    fill_in 'Email', with: 'example_user@example.com'
+    @student = FactoryGirl.create(:student)
+    @student.invite!
+    visit accept_student_invitation_path(@student, invitation_token: @student.raw_invitation_token)
     fill_in 'Password', with: 'password'
     fill_in 'Password confirmation', with: 'password'
   end
```
Comment (michaelrkn) [inline spec/features/students_pages_spec.rb]: there's no assertion in this test

File: spec/features/students_pages_spec.rb [hunk 8]
```diff
@@ -1,23 +1,25 @@
-feature 'Student signs up' do
+feature 'Student signs up via invitation' do
   before do
-    @plan = FactoryGirl.create :recurring_plan_with_upfront_payment
-    @cohort = FactoryGirl.create :future_cohort
-    visit new_student_registration_path
-    fill_in 'Email', with: 'example_user@example.com'
+    @student = FactoryGirl.create(:student)
```
Comment (michaelrkn) [inline spec/features/students_pages_spec.rb]: use `let`, not ivars

File: spec/features/admin_pages_spec.rb [hunk 9]
```diff
@@ -54,3 +55,22 @@
...
+
+feature 'Inviting new students', js: true do
+  let(:admin) { FactoryGirl.create(:admin) }
+
+  scenario 'admin selects the invite students link' do
+    login_as(admin, scope: :admin)
+    visit root_path
+    click_on 'Invite students'
+    expect(page).to have_content "Send invitation"
+  end
```
Comment (michaelrkn) [inline spec/features/admin_pages_spec.rb]: combine this test with the one below

File: spec/features/admin_pages_spec.rb [hunk 10]
```diff
@@ -14,6 +14,7 @@
     visit root_path
     expect(page).to have_link 'Assessments'
     expect(page).to have_link 'Attendance statistics'
+    expect(page).to have_link 'Invite students'
   end
```
Comment (michaelrkn) [inline spec/features/admin_pages_spec.rb]: i'm not sure we really gain anything from these tests

File: spec/features/admin_pages_spec.rb [hunk 11]
```diff
@@ -54,3 +55,22 @@
...
+  end
+
+  scenario 'when submitting an email address of a new student for an invitation email' do
+      login_as(admin, scope: :admin)
+      visit new_student_invitation_path
+      fill_in 'Email', with: 'newstudent@example.com'
+      click_on 'Send an invitation'
+      expect(page).to have_content "An invitation email has been sent to newstudent@example.com"
+  end
+end
```
Comment (michaelrkn) [inline spec/features/admin_pages_spec.rb]: add a test that students can't go to this page

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -42,6 +46,10 @@ def current_cohort
     end
   end
 
+  def authenticate_inviter!
+    authenticate_admin!(:force => true)
```
Comment (jenmccarthy) [inline app/controllers/application_controller.rb]: add comment

File: app/controllers/invitations_controller.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+class InvitationsController < Devise::InvitationsController
+
+  def after_invite_path_for(user)
+    root_path
```
Comment (jenmccarthy) [inline app/controllers/invitations_controller.rb]: add comment

File: app/models/ability.rb [hunk 14]
```diff
@@ -12,6 +12,7 @@ def initialize(user)
       can :create, Review
       can :read, CohortAttendanceStatistics
       can :create, AttendanceRecordAmendment
+      can :create, Student
```
Comment (jenmccarthy) [inline app/models/ability.rb]: try taking out and see if tests pass

File: app/models/admin.rb [hunk 15]
```diff
@@ -2,6 +2,8 @@ class Admin < User
   belongs_to :current_cohort, class_name: 'Cohort'
 
   before_create :assign_current_cohort
+  devise :database_authenticatable, :validatable, :invitable
```
Comment (jenmccarthy) [inline app/models/admin.rb]: on `User`

**Comment by jenmccarthy** [comment]
Made changes detailed above.
